internal/events/aggregator.go

```diff
@@ -509,27 +515,28 @@ func (ag *aggregator) processMessage(ctx context.Context, manifest *core.BatchMa
 		}
 
 		l.Debugf("Attempt dispatch msg=%s broadcastContexts=%v privatePins=%v", msg.Header.ID, unmaskedContexts, msg.Pins)
-		newState, dispatched, err = ag.attemptMessageDispatch(ctx, msg, data, manifest.TX.ID, state, pin)
-		if err != nil {
-			return err
-		}
+		action, correlator, err = ag.readyForDispatch(ctx, msg, data, manifest.TX.ID, state, pin)
 	}
 
-	// Mark all message pins dispatched true/false
-	// - dispatched=true: we need to write them dispatched in the DB at the end of the batch, and increment all nextPins
-	// - dispatched=false: we need to prevent dispatch of any subsequent messages on the same topic in the batch
-	if dispatched {
-		for _, np := range nextPins {
-			np.IncrementNextPin(ctx, ag.namespace)
-		}
-		state.markMessageDispatched(manifest.ID, msg, msgBaseIndex, newState)
-	} else {
+	if action == core.ActionRetry {
+		return err
+	} else if action == core.ActionWait {
+		// We need to prevent dispatch of any subsequent messages on the same topic in the batch
 		for _, unmaskedContext := range unmaskedContexts {
 			state.SetContextBlockedBy(ctx, *unmaskedContext, pin.Sequence)
 		}
+		return nil
```

Comment:
Note - there a few earlier branches of this method which return `nil` _without_ marking the context blocked. Still trying to decide if that was intentional for some reason, or if they should all fall down to here instead.